core/src/main/scala/org/scalafmt/util/TokenOps.scala

```diff
@@ -59,18 +56,32 @@ object TokenOps {
       val newlines = newlinesBetween(tok.between)
       newlines > 1 || (isDocstring(tok.right) && !tok.left.is[Comment])
     } || {
+      val leftOwner = owners(tok.left)
+      val rightOwner = owners(tok.right)
+
       style.newlines.alwaysBeforeTopLevelStatements &&
-        tok.between.count(_.is[KwNew]) < 2 && isTopLevelStatment(tok.right, owners(tok.right))
+        isTopLevelStatment(tok.right, rightOwner) &&
+        !{
+          // keep groups of case class/objects with no body in a block
+          def hasBody(tree: Tree) = tree.parent.map(_.children).getOrElse(Seq.empty).exists {
+            case Template(_, _, _, Some(_)) => true
+            case _ => false
+          }
+
+          (TreeOps.isCaseClass(leftOwner) || TreeOps.isCaseObject(leftOwner)) &&
```

Comment:
You are right, but we have at least to distingush between them since if we have 1 Block of traits and 1 Block of sealed traits we want to have newlines between the blocks
```
trait A
trait B
trait C

sealed trait ST1
sealed trait ST2
```